finality-aleph/src/data_io/data_provider.rs

```diff
@@ -0,0 +1,266 @@
+        let best_block = match &self.best_block_in_session {
+            None => {
+                // We leave the default `data_to_propose` -- we don't have the necessary input to compute a new one.
+                return;
+            }
+            Some(best_block) => best_block,
+        };
+
+        let client_info = self.client.info();
+        let finalized_block = (client_info.finalized_hash, client_info.finalized_number).into();
+
+        if let Some(prev) = &self.prev_chain_info {
+            if prev.best_block_in_session == *best_block
+                && prev.highest_finalized == finalized_block
+            {
+                // This is exactly the same state that we processed last time in update_data.
+                // No point in recomputing.
+                return;
+            }
+        }
+        // Update the info for the next call of update data.
+        self.prev_chain_info = Some(ChainInfo {
+            best_block_in_session: best_block.clone(),
+            highest_finalized: finalized_block.clone(),
+        });
+        let mut curr_block = best_block.clone();
+
+        if curr_block.num == finalized_block.num {
+            // We don't have anything to propose, we go ahead with an empty proposal.
+            *self.data_to_propose.lock().await = AlephData::Empty;
+            return;
+        }
+        if curr_block.num < finalized_block.num {
+            // Because of the client synchronization, in extremely rare cases this could happen.
+            warn!(target: "afa", "Error updating data. curr block {:?}, finalized {:?}", curr_block, finalized_block);
+            return;
+        }
+        let mut branch: Vec<B::Hash> = Vec::new();
+        while curr_block.num > finalized_block.num {
+            if curr_block.num - finalized_block.num
+                <= <NumberFor<B>>::saturated_from(MAX_DATA_BRANCH_LEN)
+            {
+                branch.push(curr_block.hash);
+            }
+            curr_block = get_parent(self.client.clone(), &curr_block)
+                .expect("block of num >= 1 must have a parent")
+        }
+        if curr_block.hash == finalized_block.hash {
+            let num_last = finalized_block.num + <NumberFor<B>>::saturated_from(branch.len());
+            // The hashes in `branch` are ordered from top to bottom -- need to reverse.
+            branch.reverse();
+            *self.data_to_propose.lock().await =
+                AlephData::HeadProposal(AlephUnvalidatedProposal::<B>::new(branch, num_last));
+        } else {
+            // By backtracking from the best block we reached a block conflicting with best finalized.
+            // This is most likely a bug, or some extremely unlikely synchronization issue of the client.
+            warn!(target: "afa", "Error updating data. Conflicting blocks: {:?}, finalized {:?}", curr_block, finalized_block);
+            // We leave the old data intact -- it is not up-to-date, but at least it is correct.
```

Comment:
This is an emergency and something went horribly wrong (our client seems to have reorganized onto a non-finalized chain). This should never happen, but if it does, it's better if we go with the old proposal instead of `Empty` -- more likely that we make progress and by some miracle recover from this failure.

Reply:
I think this is slightly less consistent than doing the opposite, but I don't think this really matters, so ok.